### ISSUE TITLE
S3 object storage issues with Linode (may affect other S3 providers?)

### ISSUE THREAD

**Issue by biglifedecision** [bookwyrm-social/bookwyrm#2323]
**Describe the bug**
assets from S3 storage are not loaded properly because of url issues.

**To Reproduce**
1. fresh install of bookwyrm on a Linode VPS running Ubuntu 22.04
2. set up S3 object storage using Linode's S3 Object Storage buckets, following bookwyrm's External Storage instructions: https://docs.joinbookwyrm.com/external-storage.html

example settings in .env:
```
# S3 configuration
USE_S3=true
AWS_ACCESS_KEY_ID=$$$
AWS_SECRET_ACCESS_KEY=$$$

AWS_STORAGE_BUCKET_NAME="mybookwyrmfiles"
AWS_S3_CUSTOM_DOMAIN="mybookwyrmfiles.us-east-1.linodeobjects.com"
AWS_S3_REGION_NAME="us-east-1"
AWS_S3_ENDPOINT_URL="https://mybookwyrmfiles.us-east-1.linodeobjects.com"
```

Notably:
3. run `./bd-dev collectstatic` with success
4. run `./bd-dev up` with success
5. navigate to bookwyrm in my browser: none of the S3 assets load, so there are no images, the CSS isn't loaded, etc. 

**Expected behavior**
The `collectstatic` command is putting all the static assets into the S3 storage. Based on the URL where bookwyrm's HTML is trying to fetch those assets, it should be putting them at the Bucket root folder / (under a static subfolder)

The bookwyrm page source shows that, for example, it is trying to fetch the CSS file from: 
https://mybookwyrmfiles.us-east-1.linodeobjects.com/static/css/themes/bookwyrm-light.css

However, I can look in my Object Storage Bucket on the Linode dashboard and see that all the assets are actually in a folder /mybookwyrmfiles/ - a subfolder the same name as the bucket.

So the CSS file is actually here: 
https://mybookwyrmfiles.us-east-1.linodeobjects.com/mybookwyrmfiles/static/css/themes/bookwyrm-light.css

I can easily load it directly at this URL in my browser, so I think it is not a CORS/permission/access issue.

Trying in to change .env to something like: `AWS_S3_ENDPOINT_URL="https://mybookwyrmfiles.us-east-1.linodeobjects.com/mybookwyrmfiles"` does not appear to change the behaviour of the bookwyrm instance. No matter what I do, loading bookwyrm in the browser it just tries to fetch at https://mybookwyrmfiles.us-east-1.linodeobjects.com/static/ even though the files are not actually there. 

If I modify `AWS_S3_ENDPOINT_URL` and run `collectstatic` again, it will stick the files in whatever subfolder indicated by the path, but it will always be that one extra subfolder down. And regardless, bookwyrm itself is always going to the bucket-url/static path to look for these files, no matter what `AWS_S3_ENDPOINT_URL` is set to. Am I missing a step to tell bookwyrm to look at a new endpoint url for S3 stored assets? Or am I really lost?

eg
`AWS_S3_ENDPOINT_URL="https://mybookwyrmfiles.us-east-1.linodeobjects.com/nest/nest/"`
will put the files at https://mybookwyrmfiles.us-east-1.linodeobjects.com/mybookwyrmfiles/nest/nest

I'm not an expert on S3 stuff by any means, so I am not sure what expected behaviour should be. Changing `AWS_S3_ENDPOINT_URL` will change where `collectstatic` puts the files, but has no effect on where bookwyrm's HTML ultimately tries to fetch the files. and bookwyrm's HTML is unable to fetch the files in anycase, because they are always placed one folder down - this issue I am guessing is something to do with Linode's implementation of S3?

I have not tried actually uploading any user files yet. I may try indicating that i want the static assets on the server itself instead of in the S3 bucket, and then see if uploads work. 

Also very likely this is all just me missing something about S3, or just a basic thing. This is my first time installing something with docker, second time dealing with S3 and probably the 4th or 5th thing like this I have ever installed at all! So hoping if it's more me I can at least help improve the instructions based on my failure!

**Comment by mouse-reeve** [comment]
Huh! I would really have expected using `AWS_S3_ENDPOINT_URL="https://mybookwyrmfiles.us-east-1.linodeobjects.com/mybookwyrmfiles"` to fix the problem. When you made that change, is it possible that the image didn't properly collect the updated .env file? As I learned the hard way, `docker-compose restart` does _not_ pick up environment changes. `docker-compose up`, however, should

**Comment by biglifedecision** [comment]
That was also what I expected!

It is possible that the updated .env file isn't getting 'collected' as you say. I'm a total docker novice. I am just following the bookwyrm instructions so i definitely could be missing something. after modifying .env I run docker-compose up again.

In my troubleshooting, I was able to turn off S3 storage, then run `collectstatic` (so the static assets are placed locally) then `docker-compose up` and then bookwyrm appears to properly load all the assets locally. so it seems to be reading the change in .env to turn S3 storage on/off. but there could be a bug so that the url_endpoint update doesn't get picked up? that's what i was thinking.

Thank you for your help!

**Comment by biglifedecision** [comment]
Are there any commands I could try to get the `AWS_S3_ENDPOINT_URL` to pick up from .env? Or to tinker with `collectstatic` to change where it puts the files on the S3 bucket?

**Comment by mouse-reeve** [comment]
The collectstatic commands for s3 are installed, not part of bookwyrm directly. I'm going to see if anyone in the developer chat has any ideas!

**Comment by damian-murphy** [comment]
Hi, I've precisely the same problem! Can't see how it's happening, looked through the bw-dev commands, the storage_backends dor django, all looks right, but somehow it's all been copied to my linode bucket under a key starting with the bucket name, breaking all /images/ and /static/ file links :-/

Did you get any suggestions?

Thanks

**Comment by damian-murphy** [comment]
After some debugging, I've worked out what's happening, and it's specific here to Linode S3. The catch was that by specifying the bucket name in the endpoint URL and DNS, it ends up appending the bucket name to the storage key as a prefix. So, for example, this works:

AWS_STORAGE_BUCKET_NAME=my_bucket_name
AWS_S3_CUSTOM_DOMAIN=linode-region-name.linodeobjects.com
AWS_S3_REGION_NAME=linode-region-name
AWS_S3_ENDPOINT_URL=https://linode-region-name.linodeobjects.com

So, even though the URL you're given in the linode admin contains the bucket name, only put it in the AWS_STORAGE_BUCKET_NAME and leave it out of any other domain or URL.

**Comment by biglifedecision** [comment]
@damian-murphy so it worked? i thought i tried every possible debugging scenario without luck, but i've also been meaning to try again since bookwyrm has had several updates since... so maybe i missed this, or maybe something's changed. i'm going to see if i can replicate your results, thanks for sharing this!

**Comment by damian-murphy** [comment]
@biglifedecision you're welcome! Seems to have worked, at least the bucket objects look correct now. I'm waiting on the copy task to finish, then I can actually confirm.

**Comment by damian-murphy** [comment]
@biglifedecision I spoke too soon! The URL's it's trying to reach the objects on is now wrong. It's missing the bucket name on the link url - it just has the cluster-id.linodeobjects.com address. Close, but not close enough.

**Comment by biglifedecision** [comment]
ok :(

**Comment by damian-murphy** [comment]
@biglifedecision Right, I seem to have a 'fix'. At least, my instance is now operational. I'll file a documentation update with this info.

So, the issue seems to be that for Linode (only one I'm testing with), you need to do the following ('worked for me'):

If you are migrating existing content, then:
- Set the AWS_ vars as per comment https://github.com/bookwyrm-social/bookwyrm/issues/2323#issuecomment-1812712418
- Run the steps as per documentation https://docs.joinbookwyrm.com/external-storage.html, specifically:
 - ./bw-dev copy_media_to_s3
 - ./bw-dev compile_themes
 - ./bw-dev collectstatic
 - Do the CORS, set HTTPS, set USE_S3 and all that, and restart your instance.

You'll have migrated your content (you can see it in your bucket in linode, in the root under /images and /static

BUT!
Your instance links will be broken - no css, no icons, no images.

To fix this, edit the variable '`AWS_S3_CUSTOM_DOMAIN`' to add your bucket name to it, e.g. if your bucket is called `bookwyrm-store` and your AWS_S3_CUSTOM_DOMAIN is `eu-central-1.linodeobjects.com`, then it becomes `bookwyrm-store.eu-central-1.linodeobjects.com`

For some reason, this works perfectly. New content is sent to the object store (I tested this). Links work. But, if you try to re-sync or copy content again, the copy process is broken and will create objects in the root of your bucket prefixed with the bucket name!

Obviously, if you're NOT migrating content and starting fresh, then you might have an easier time of it, although I suspect this will still need to be done for the static content move.

**Comment by biglifedecision** [comment]
wow. thank you!

**Comment by hughrun** [comment]
This feels pretty hacky but since this is specific to Linode, and we have updated documentation, I'm going to close this issue.
